Include animations and transitions in the cascade

mrobinson opened the pull request:
Instead of applying animations and transitions to styled elements,
include them in the cascade. This allows them to interact properly with
things like font-size and !important rules.

<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

mrobinson committed:
Include animations and transitions in the cascade

Instead of applying animations and transitions to styled elements,
include them in the cascade. This allows them to interact properly with
things like font-size and !important rules.

highfive commented:
Heads up! This PR modifies the following files:
 * @emilio: components/style/sharing/mod.rs, components/style/traversal.rs, components/style/style_resolver.rs, components/style/properties/properties.mako.rs, components/style/dom.rs and 3 more

emilio commented on components/style/matching.rs:
```diff
@@ -493,6 +528,20 @@ trait PrivateMatchMethods: TElement {
+                RuleInclusion::All,
+                PseudoElementResolution::IfApplicable,
+            );
+            let new_primary = resolver.resolve_style_with_default_parents();
```
I'm a bit confused about this. Why does this need to `resolve_*`, then re-cascade?

mrobinson replied:
You're right. Re-cascading here is bogus. I've fixed this in the latest version of the branch.

emilio replied:
So... I'm still a bit unclear on why this is the right thing to do. This is not just recascading but also re-selector-matching. I'd expect this to do basically just [this](https://searchfox.org/mozilla-central/rev/9ad88f80aeedcd3cd7d7f63be07f577861727054/servo/components/style/traversal.rs#622-635).

And to be clear, I think this is fine as a stepping-stone if needed, but I don't understand why something like that wouldn't work.

mrobinson replied:
You're right. I think I was trying this before, but not properly replacing rules. I'll need to do it a bit more carefully because `replace_rules` is written with the animations-only restyle in mind. Thanks for the feedback.

mrobinson commented:
@emilio, thanks for taking a look at this. I've addressed your comment and also have made a few changes in order to avoid re-resolving the style when animations or transitions aren't changing.

mrobinson committed:
animations: Don't always re-resolve the node style

When animations and transitions change don't always re-resolve node
style, just replace the animation and transition rules and re-cascade.

mrobinson commented:
I've pushed a new commit to the branch that replaces the transition and animations rules directly and then does a re-cascade.

mrobinson commented:
I have a pushed a new version of the branch that uses `StyleResolver::.cascade_style_and_visited_with_default_parents` instead of `cascade_styles_with_default_parents`, simplifying the second commit.

emilio commented on components/style/animation.rs:
```diff
@@ -875,17 +864,30 @@ pub struct ElementAnimationSet {
+    }
+
+    fn cancel_active_transitions(&mut self) {
+        self.dirty = !self.transitions.is_empty();
```
Does this dirty too much? Should it just dirty right before the `transition.state = ` assignment?

emilio commented on components/style/matching.rs:
```diff
@@ -458,10 +465,60 @@ trait PrivateMatchMethods: TElement {
+        }
+
+        // If we have modified animation or transitions, we recascade style for this node.
+        let mut rule_node = new_resolved_styles.primary.style.0.rules().clone();
```
May be worth exposing a `ResolvedElementStyles::primary_style()` or something so that this reads a bit better, but no big deal.

emilio commented on components/style/matching.rs:
```diff
@@ -458,10 +465,60 @@ trait PrivateMatchMethods: TElement {
+                .as_ref()
+                .map(|a| a.borrow_arc()),
+            &mut rule_node,
+        );
```
May be worth either asserting that `rule_node` is not the primary style's rule node anymore, or check for that and bail out.

emilio reviewed (approved):
Thank you, this looks pretty sensible!
